### PR TITLE
Issue #3142: Several updates for the custom certificate mapper API:

### PR DESCRIPTION
**Pull request by jvanhill**
1) Renamed CertificateMapper to X509CertificateMapper
2) Changed mapCertificate(..) method to take a certificate chain instead of a single certificate.
3) Leverage change in BELLs ability to now set OSGi service properties. Can set the x509.certificate.mappper.id to the X509CertificateMapper and reference it from the server configuration.

### CODE REVIEW

File: dev/com.ibm.ws.security.wim.adapter.ldap/resources/com/ibm/ws/security/wim/adapter/ldap/resources/LdapUtilMessages.nlsprops [hunk 1]
```diff
@@ -204,4 +204,20 @@ ENTITY_ALREADY_EXIST.useraction=Retry the operation and specify a different name
...
+LDAP_REGISTRY_CUSTOM_MAPPER_NOT_SUPPORTED.explanation=The custom X.509 certificate mapper threw a CertificateMapNotSupportedException.
+LDAP_REGISTRY_CUSTOM_MAPPER_NOT_SUPPORTED.useraction=If the exception was unexpected, review the custom X.509 certificate mapper implementation and make necessary changes.
+
+LDAP_REGISTRY_CUSTOM_MAPPER_FAILED=CWIML4503E: The custom X.509 certificate mapper threw a CertificateMapFailedException.
+LDAP_REGISTRY_CUSTOM_MAPPER_FAILED.explanation=The custom X.509 certificate mapper threw a CertificateMapFailedException.
+LDAP_REGISTRY_CUSTOM_MAPPER_FAILED.useraction=If the exception was unexpected, review the custom X.509 certificate mapper implementation and make necessary changes.
+
+LDAP_REGISTRY_INVALID_MAPPING=CWIML4504W: The custom X.509 certificate mapper returned an empty or null value. 
+LDAP_REGISTRY_INVALID_MAPPING.explanation=The custom X.509 certificate mapper cannot return an empty or null value.
+LDAP_REGISTRY_INVALID_MAPPING.useraction=Ensure that the custom X.509 certificate mapper implementation does not return an empty or null value. When a valid value cannot be returned an exception should be thrown instead.
```
Comment (sishida) [inline dev/com.ibm.ws.security.wim.adapter.ldap/resources/com/ibm/ws/security/wim/adapter/ldap/resources/LdapUtilMessages.nlsprops]: What does the user need to do to ensure that an empty or null value isn't returned? How can they make the implementation throw an exception instead?

Also, there should be a comma after "returned" in, "When a valid value cannot be returned, an exception should be thrown instead."

File: dev/com.ibm.ws.security.registry.basic/resources/com/ibm/ws/security/registry/basic/internal/resources/LoggingMessages.nlsprops [hunk 2]
```diff
@@ -53,3 +53,27 @@ BASIC_REGISTRY_SAME_MEMBER_DEFINITION.useraction=Define only one member entry fo
 BASIC_REGISTRY_UNKNOWN_MEMBER_DEFINITION=CWWKS3107W: Member entry with the name ''{0}'' for group ''{1}'' does not match a defined user.
 BASIC_REGISTRY_UNKNOWN_MEMBER_DEFINITION.explanation=The name specified for the member entry does not match any defined users.
 BASIC_REGISTRY_UNKNOWN_MEMBER_DEFINITION.useraction=Check the spelling of the member name and ensure it matches a user defined in the registry.
+
+BASIC_REGISTRY_MAPPER_NOT_BOUND=CWWKS3108W: A custom X.509 certificate mapper was not bound with this basic registry.
+BASIC_REGISTRY_MAPPER_NOT_BOUND.explanation=The X.509 certificate mapping mode was set to CUSTOM, but no custom X.509 certificate mapper was bound to the basic registry. 
+BASIC_REGISTRY_MAPPER_NOT_BOUND.useraction=Ensure that the basic registry's ''certificateMapperId'' attribute is configured to point to a valid custom X.509 certificate mapper.
```
Comment (sishida) [inline dev/com.ibm.ws.security.registry.basic/resources/com/ibm/ws/security/registry/basic/internal/resources/LoggingMessages.nlsprops]: "Ensure that the certificateMapperId attribute in the basic registry is..."

File: dev/com.ibm.ws.security.registry.basic/resources/com/ibm/ws/security/registry/basic/internal/resources/LoggingMessages.nlsprops [hunk 3]
```diff
@@ -53,3 +53,27 @@ BASIC_REGISTRY_SAME_MEMBER_DEFINITION.useraction=Define only one member entry fo
...
+BASIC_REGISTRY_CUSTOM_MAPPER_NOT_SUPPORTED=CWWKS3110W: The custom X.509 certificate mapper threw a CertificateMapNotSupportedException.
+BASIC_REGISTRY_CUSTOM_MAPPER_NOT_SUPPORTED.explanation=The custom X.509 certificate mapper threw a CertificateMapNotSupportedException.
+BASIC_REGISTRY_CUSTOM_MAPPER_NOT_SUPPORTED.useraction=If the exception was unexpected, review the custom X.509 certificate mapper implementation and make necessary changes.
+
+BASIC_REGISTRY_CUSTOM_MAPPER_FAILED=CWWKS3111E: The custom X.509 certificate mapper threw a CertificateMapFailedException.
+BASIC_REGISTRY_CUSTOM_MAPPER_FAILED.explanation=The custom X.509 certificate mapper threw a CertificateMapFailedException.
+BASIC_REGISTRY_CUSTOM_MAPPER_FAILED.useraction=If the exception was unexpected, review the custom X.509 certificate mapper implementation and make necessary changes.
+
+BASIC_REGISTRY_NAME_NOT_FOUND=CWWKS3112W: The distinguished name (DN) ''{0}'' does not map to a valid basic registry user.
+BASIC_REGISTRY_NAME_NOT_FOUND.explanation=The subject principal's distinguished name (DN) in the X.509 certificate does not map to a valid basic registry user.
```
[the remark this answers is not in the document]
Comment (sishida) [reply dev/com.ibm.ws.security.registry.basic/resources/com/ibm/ws/security/registry/basic/internal/resources/LoggingMessages.nlsprops]: "The DN of the subject principal..."

File: dev/com.ibm.ws.security.registry.basic/resources/com/ibm/ws/security/registry/basic/internal/resources/LoggingMessages.nlsprops [hunk 4]
```diff
@@ -53,3 +53,27 @@ BASIC_REGISTRY_SAME_MEMBER_DEFINITION.useraction=Define only one member entry fo
...
+BASIC_REGISTRY_CUSTOM_MAPPER_FAILED.useraction=If the exception was unexpected, review the custom X.509 certificate mapper implementation and make necessary changes.
+
+BASIC_REGISTRY_NAME_NOT_FOUND=CWWKS3112W: The distinguished name (DN) ''{0}'' does not map to a valid basic registry user.
+BASIC_REGISTRY_NAME_NOT_FOUND.explanation=The subject principal's distinguished name (DN) in the X.509 certificate does not map to a valid basic registry user. 
+BASIC_REGISTRY_NAME_NOT_FOUND.useraction=Ensure that the subject principal's distinguished name (DN) contains a 'cn' relative distinguished name (RDN) and that there is a user in the basic registry that matches the RDN's value. 
+
+BASIC_REGISTRY_CERT_IGNORED=CWWKS3113W: The basic registry is configured to ignore X.509 certificate authentication requests.
+BASIC_REGISTRY_CERT_IGNORED.explanation=The 'certificateMapMode' configuration attribute for the basic registry is set to 'IGNORE'.
+BASIC_REGISTRY_CERT_IGNORED.useraction=If the basic registry should process X.509 certificate authentication requests the 'certificateMapMode' configuration attribute for the registry should be changed to allow processing.
```
Comment (sishida) [inline dev/com.ibm.ws.security.registry.basic/resources/com/ibm/ws/security/registry/basic/internal/resources/LoggingMessages.nlsprops]: To enable the basic registry to process X.509 certificate authentication requests, change the 'certificateMapMode' configuration attribute to allow processing.

File: dev/com.ibm.ws.security.registry.basic/resources/com/ibm/ws/security/registry/basic/internal/resources/LoggingMessages.nlsprops [hunk 5]
```diff
@@ -53,3 +53,27 @@ BASIC_REGISTRY_SAME_MEMBER_DEFINITION.useraction=Define only one member entry fo
...
 BASIC_REGISTRY_UNKNOWN_MEMBER_DEFINITION.explanation=The name specified for the member entry does not match any defined users.
 BASIC_REGISTRY_UNKNOWN_MEMBER_DEFINITION.useraction=Check the spelling of the member name and ensure it matches a user defined in the registry.
+
+BASIC_REGISTRY_MAPPER_NOT_BOUND=CWWKS3108W: A custom X.509 certificate mapper was not bound with this basic registry.
+BASIC_REGISTRY_MAPPER_NOT_BOUND.explanation=The X.509 certificate mapping mode was set to CUSTOM, but no custom X.509 certificate mapper was bound to the basic registry. 
+BASIC_REGISTRY_MAPPER_NOT_BOUND.useraction=Ensure that the basic registry's ''certificateMapperId'' attribute is configured to point to a valid custom X.509 certificate mapper.
+
+BASIC_REGISTRY_MAPPED_NAME_NOT_FOUND=CWWKS3109W: The filter or distinguished name ''{0}'' returned from the X.509 custom certificate mapper does not map to a registry user.
+BASIC_REGISTRY_MAPPED_NAME_NOT_FOUND.explanation=The value returned from the custom X.509 certificate mapper does not map to a user in the basic registry.
```
Comment (sishida) [inline dev/com.ibm.ws.security.registry.basic/resources/com/ibm/ws/security/registry/basic/internal/resources/LoggingMessages.nlsprops]: "The value that is returned..."

**Comment by sishida** [comment]
I added comments for ID review.

Two comments that would apply across all the messages:
1. Remove "quotation marks" where possible.
2. Make sentences present tense wherever possible, such as changing "was" to "is."

Thank you!

**Comment by jvanhill** [comment]
@sishida - I have updated the messages. Please take a look.

**Comment by jvanhill** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.security.wim.adapter.file_fat,com.ibm.ws.security.wim.adapter.ldap_fat,com.ibm.ws.security.wim.core_fat,com.ibm.ws.security.wim.registry_fat,com.ibm.ws.security.wim.scim_fat

**Comment by jvanhill** [comment]
#build

**Comment by jvanhill** [comment]
Re-building since an LDAP FAT failed, but it appears to not be related to my changes.

#build

File: dev/com.ibm.ws.security.registry.basic/resources/OSGI-INF/l10n/metatype.properties [hunk 6]
```diff
@@ -50,11 +50,11 @@ member.name.desc=Name of a user in a Basic User Registry group.
...
+certificate.mapper.id=Certificate mapper ID
+certificate.mapper.id.desc=Specifies the X509CertificateMapper to use when the X.509 certificate authentication mapping mode is CUSTOM. The value must match the value of the 'x509.certificate.mapper.id' property that is specified for the X509CertificateMapper implementation.
 
 certificate.map.mode.principal_cn=The basic registry attempts to authenticate the X.509 certificate by mapping the PrincipalName value in the X.509 certificate to the exact distinguished name (DN) in the repository. If a single matching entity is found, the authentication is successful. If a matching entity is not found or more than a single matching entity is found, authentication fails and the program returns an error.
-certificate.map.mode.custom=The basic registry attempts to authenticate the X.509 certificate by using the custom CertificateMapper implementation specified by certificateMapperId. If a single matching entity is found, the authentication is successful. If a matching entity is not found or more than a single matching entity is found, authentication fails and the program returns an error.
```
Comment (sishida) [inline dev/com.ibm.ws.security.registry.basic/resources/OSGI-INF/l10n/metatype.properties]: More commas: If a matching entity is not found**,** or more than a single matching entity is found, authentication fails**,** and the program returns an error.

File: dev/com.ibm.ws.security.registry.basic/resources/OSGI-INF/l10n/metatype.properties [hunk 7]
```diff
@@ -50,11 +50,11 @@ member.name.desc=Name of a user in a Basic User Registry group.
...
-certificate.map.mode.custom=The basic registry attempts to authenticate the X.509 certificate by using the custom CertificateMapper implementation specified by certificateMapperId. If a single matching entity is found, the authentication is successful. If a matching entity is not found or more than a single matching entity is found, authentication fails and the program returns an error. 
-
+certificate.map.mode.custom=The basic registry attempts to authenticate the X.509 certificate by using the custom X509CertificateMapper implementation specified by certificateMapperId. If a single matching entity is found, the authentication is successful. If a matching entity is not found or more than a single matching entity is found, authentication fails and the program returns an error.
```
Comment (sishida) [inline dev/com.ibm.ws.security.registry.basic/resources/OSGI-INF/l10n/metatype.properties]: "implementation **that is** specified by..."
Give certificateMapperId a label word after it like you did with **the** X.509 **certificate**.

File: dev/com.ibm.ws.security.registry.basic/resources/OSGI-INF/l10n/metatype.properties [hunk 8]
```diff
@@ -50,11 +50,11 @@ member.name.desc=Name of a user in a Basic User Registry group.
...
-certificate.map.mode.custom=The basic registry attempts to authenticate the X.509 certificate by using the custom CertificateMapper implementation specified by certificateMapperId. If a single matching entity is found, the authentication is successful. If a matching entity is not found or more than a single matching entity is found, authentication fails and the program returns an error. 
-
+certificate.map.mode.custom=The basic registry attempts to authenticate the X.509 certificate by using the custom X509CertificateMapper implementation specified by certificateMapperId. If a single matching entity is found, the authentication is successful. If a matching entity is not found or more than a single matching entity is found, authentication fails and the program returns an error. 
+certificate.map.mode.not.supported=The LDAP registry does not support X.509 certificate authentication. Attempts to authenticate with an X.509 certificate will fail and a CertificateMapNotSupportedException exception will be thrown.
```
Comment (sishida) [inline dev/com.ibm.ws.security.registry.basic/resources/OSGI-INF/l10n/metatype.properties]: Use present tense it appropriate in context.
Add comma: "...fail**,** and..."

File: dev/com.ibm.ws.security.wim.adapter.ldap/resources/com/ibm/ws/security/wim/adapter/ldap/resources/LdapUtilMessages.nlsprops [hunk 9]
```diff
@@ -204,4 +204,20 @@ ENTITY_ALREADY_EXIST.useraction=Retry the operation and specify a different name
...
+LDAP_REGISTRY_CUSTOM_MAPPER_NOT_SUPPORTED.useraction=If the exception was unexpected, review the custom X.509 certificate mapper implementation and make necessary changes.
+
+LDAP_REGISTRY_CUSTOM_MAPPER_FAILED=CWIML4503E: The custom X.509 certificate mapper threw a CertificateMapFailedException exception.
+LDAP_REGISTRY_CUSTOM_MAPPER_FAILED.explanation=The custom X.509 certificate mapper threw a CertificateMapFailedException exception.
+LDAP_REGISTRY_CUSTOM_MAPPER_FAILED.useraction=If the exception was unexpected, review the custom X.509 certificate mapper implementation and make necessary changes.
+
+LDAP_REGISTRY_INVALID_MAPPING=CWIML4504W: The custom X.509 certificate mapper returned an empty or null value. 
+LDAP_REGISTRY_INVALID_MAPPING.explanation=The custom X.509 certificate mapper must return a non-empty and non-null value or throw one of the named exceptions.
+LDAP_REGISTRY_INVALID_MAPPING.useraction=Modify the custom X.509 certificate mapper implementation to not return an empty or null value. When a valid value cannot be returned, a checked exception should be thrown instead.
```
Comment (sishida) [inline dev/com.ibm.ws.security.wim.adapter.ldap/resources/com/ibm/ws/security/wim/adapter/ldap/resources/LdapUtilMessages.nlsprops]: Could we change to, "...a checked exception **is** thrown instead."?

**Comment by jvanhill** [comment]
#build

**Comment by jvanhill** [comment]
The build test failures were do to a regression commited prior to my build and is unrelated to my changes.

**Pull request merged**